### ISSUE TITLE
IIFE closure extension temporary value

### ISSUE THREAD

**Issue by A4-Tacks** [rust-lang/rust-clippy#15290]
### Summary

Use IIFE closure extension temporary value (`format_args`)

### Lint Name

redundant_closure_call

### Reproducer

I tried this code:

```rust
fn main() {
    (|args| {
        println!("{args}");
    })(format_args!("test: {}", "foo".to_owned()))
}
```

I saw this happen:

```
warning: try not to call a closure in the expression where it is declared
 --> src/bin/closure-format_args.rs:2:5
  |
2 | /     (|args| {
3 | |         println!("{args}");
4 | |     })(format_args!("test: {}", "foo".to_owned()))
  | |__________________________________________________^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_ca
ll
  = note: `#[warn(clippy::redundant_closure_call)]` on by default

warning: `quick-rs` (bin "closure-format_args") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.36s
```

I expected to see this happen:

No this warning

### Version

```text
rustc 1.90.0-nightly (a00149764 2025-07-14)
binary: rustc
commit-hash: a001497644bc229f1abcc5b2528733386591647f
commit-date: 2025-07-14
host: aarch64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.8
```

**Comment by blyxyas** [comment]
Why is this a false positive? This linting case is about calling a closure in the same expression that is defined (i.e. any `(|..| ..)(..)`)

This case also lints:

```rust
fn main() {
    (|args| {
        args;
    })(1);
}
```

**Comment by Jarcho** [comment]
You're example works without a closure.

```rust
fn main() {
    println!("{}", format_args!("test: {}", "foo".to_owned()))
}
```

**Comment by A4-Tacks** [comment]
```rust
fn side_effect() -> i32 { 2 }
fn main() {
    (|args| {
        let n = side_effect();
        println!("{args}+{n}");
    })(format_args!("test: {}", "foo".to_owned()))
}
```
Change to:
```rust
fn side_effect() -> i32 { 2 }
fn main() {
    {
        let args = format_args!("test: {}", "foo".to_owned()); // error
        let n = side_effect();
        println!("{args}+{n}");
    }
}
```

Expexted `"foo".to_owned()` before of `side_effect()`, do not inline it

**Comment by ada4a** [comment]
> Expexted `"foo".to_owned()` before of `side_effect()`, do not inline it

I think `"foo".to_owned()` happens _after_ `side_effect()` in both cases. To see that more clearly, consider the following example:
```rs
struct PrintOnCreation(i32);
impl PrintOnCreation {
    fn new(n: i32) -> Self {
        eprintln!("{n}");
        Self(n)
    }
}

fn main() {
    (|p| {
        PrintOnCreation::new(2);
    })(PrintOnCreation::new(1));
}
```
If you run this, you'll get the following output:
```
1
2
```
so `PrintOnCreation::new(1)` is called first, even if it's located further down than `PrintOnCreation::new(2);` in the source code.

That's because a closure isn't run when it's created, only when it's called. But before it can be run, its arguments (in this case, `PrintOnCreation::new(1)`) need to be evaluated first.
